pennylane/measurements/shots.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the Shots class to hold shot-related information."""
+# pylint:disable=inconsistent-return-statements
+from collections import namedtuple
+from typing import Sequence, Tuple
+
+ShotCopies = namedtuple("ShotCopies", ["shots", "copies"])
+"""A namedtuple that represents a shot quantity being repeated some number of times."""
+
+
+def valid_int(s):
+    """Returns True if s is a positive integer."""
+    return isinstance(s, int) and s > 0
+
+
+def valid_tuple(s):
+    """Returns True if s is a tuple of the form (shots, copies)."""
+    return isinstance(s, tuple) and len(s) == 2 and valid_int(s[0]) and valid_int(s[1])
+
+
+class Shots:
+    """
+    A data class that stores shot information.
+
+    Args:
+        shots (Union[None, int, Tuple[int, int], Sequence[int, Tuple[int, int]]]): Raw shot information
```

Comment:
I think since this is so heavily dispatched, we should consider a different format for the documentation.  Maybe we can mirror how `qml.evolve` does it, where each type of argument gets its own heading.

Reply:
true, but those various types are subclasses of each other so they still give the `Operator` type-hint. Anyway, we can remove `Tuple[int, int]` from here given what you've described, so it won't be toooo bad?